trie/zk_trie_impl.go

```diff
@@ -162,157 +176,176 @@ func (mt *ZkTrieImpl) pushLeaf(newLeaf *Node, oldLeaf *Node, lvl int,
 		if err != nil {
-			//fmt.Printf("addLeaf: GetNode err %v level %v\n", err, lvl)
 			return nil, false, err
 		}
-		// Update the node to reflect the modified child
-		hash, err := mt.addNode(newParentNode)
-		return hash, false, err
+
+		// do nothing, if child subtrie was not modified
+		if newChildSubTrieRoot == nil {
+			return nil, false, nil
+		}
+
+		newNodetype := n.Type
+		if !isTerminal {
+			newNodetype = newNodetype.DeduceUpgradeType(branchRight)
+		}
+
+		var newNode *Node
+		if branchRight {
+			newNode = NewParentNode(newNodetype, n.ChildL, newChildSubTrieRoot)
+		} else {
+			newNode = NewParentNode(newNodetype, newChildSubTrieRoot, n.ChildR)
+		}
+
+		// if current node is already dirty, modify in-place
+		// else create a new dirty sub-trie
+		newCurTrieRootKey := mt.newDirtyNodeKey()
+		mt.dirtyStorage[*newCurTrieRootKey] = newNode
+		return newCurTrieRootKey, false, err
 	case NodeTypeEmpty, NodeTypeLeaf, NodeTypeParent:
 		panic("encounter unsupported deprecated node type")
 	default:
 		return nil, false, ErrInvalidNodeFound
 	}
 }
 
-// addNode adds a node into the MT and returns the node hash. Empty nodes are
-// not stored in the tree since they are all the same and assumed to always exist.
-func (mt *ZkTrieImpl) addNode(n *Node) (*zkt.Hash, error) {
-	// verify that the ZkTrieImpl is writable
-	if !mt.writable {
-		return nil, ErrNotWritable
-	}
-	if n.Type == NodeTypeEmpty {
-		return n.NodeHash()
+// newDirtyNodeKey increments the dirtyIndex and creates a new dirty node key
+func (mt *ZkTrieImpl) newDirtyNodeKey() *zkt.Hash {
+	mt.dirtyIndex.Add(mt.dirtyIndex, zkt.BigOne)
+	return zkt.NewHashFromBigInt(mt.dirtyIndex)
+}
+
+// isDirtyNode returns if the node with the given key is dirty or not
+func (mt *ZkTrieImpl) isDirtyNode(nodeKey *zkt.Hash) bool {
+	_, found := mt.dirtyStorage[*nodeKey]
+	return found
+}
+
+// commit calculates the commitment for the given sub trie and persists all dirty nodes
+func (mt *ZkTrieImpl) commit(rootKey *zkt.Hash, hashedDirtyNodes map[zkt.Hash]*Node, commitLock *sync.Mutex) (*zkt.Hash, error) {
+	if !mt.isDirtyNode(rootKey) {
+		return rootKey, nil
 	}
-	hash, err := n.NodeHash()
+
+	root, err := mt.GetNode(rootKey)
 	if err != nil {
 		return nil, err
 	}
-	v := n.CanonicalValue()
-	// Check that the node key doesn't already exist
-	oldV, err := mt.db.Get(hash[:])
-	if err == nil {
-		if !bytes.Equal(oldV, v) {
-			//fmt.Printf("Encounter conflicted node hash: %x, old value %x and new %x\n", hash, oldV, v)
-			return nil, ErrNodeKeyAlreadyExists
-		} else {
-			// duplicated
-			return hash, nil
+
+	switch root.Type {
+	case NodeTypeEmpty:
+		return &zkt.HashZero, nil
+	case NodeTypeLeaf_New:
+		// leaves are already hashed, we just need to persist it
+		break
+	case NodeTypeBranch_0, NodeTypeBranch_1, NodeTypeBranch_2, NodeTypeBranch_3:
+		leftDone := make(chan struct{})
+		var leftErr error
+		go func() {
```

Comment:
I too considered limiting the concurrency here, but decided it probably won't going to be an issue in practice. Mainly because our blocks are pretty small for the time being, because of CCC, so the change set on the trie is fairly small as well. But I can definitely limit the concurrency here, if you feel like that is needed.